sa/sa.go

```diff
@@ -915,3 +915,188 @@ func (ssa *SQLStorageAuthority) Health(ctx context.Context) error {
 	}
 	return nil
 }
+
+// LeaseCRLShard marks a single crlShards row as leased until the given time.
+// If the request names a specific shard, this function will return an error
+// if that shard is already leased. Otherwise, this function will return the
+// index of the oldest shard for the given issuer.
+func (ssa *SQLStorageAuthority) LeaseCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
+	if core.IsAnyNilOrZero(req.Until, req.IssuerNameID) {
+		return nil, errIncompleteRequest
+	}
+	if req.Until.AsTime().Before(ssa.clk.Now()) {
+		return nil, fmt.Errorf("lease timestamp must be in the future, got %q", req.Until.AsTime())
+	}
+
+	if req.MinShardIdx == req.MaxShardIdx {
+		return ssa.leaseSpecificCRLShard(ctx, req)
+	} else {
+		return ssa.leaseOldestCRLShard(ctx, req)
+	}
+}
+
+// leaseOldestCRLShard finds and then leases the crl shard for the given issuer
+// which is not currently leased and hasn't been updated in the longest. It
+// returns an error if all shards for that issuer are already leased.
+func (ssa *SQLStorageAuthority) leaseOldestCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
+	shardIdx, err := db.WithTransaction(ctx, ssa.dbMap, func(txWithCtx db.Executor) (interface{}, error) {
+		var shards []*crlShardModel
+		_, err := txWithCtx.Select(
+			&shards,
+			`SELECT id, issuerID, idx, thisUpdate, nextUpdate, leasedUntil
+			FROM crlShards
+			WHERE issuerID = ?`,
+			req.IssuerNameID,
+		)
+		if err != nil {
+			return -1, err
+		}
+
+		// Convert the slice to a map, so we can easily detect shards that we expect
+		// to exist, but don't. At the same time, keep track of the oldest shard
+		// we've seen, so we can work on that one if none are missing.
+		shardMap := make(map[int]*crlShardModel, len(shards))
```

Comment:
Since we know how many shards and the number of indexes per shard, we could enforce the capacity too.

Reply:
That's actually what this second argument is doing: you can't provide a starting length for a map (it has exactly as many items as it has keys, end of story), so the optional argument to `make` supplies a capacity instead.